[BEAM-1000,BEAM-595,BEAM-1050] Aligned the Spark batch and streaming pipelines to adhere to waitUntilFinish and cancel

staslev opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

Introduced SparkPipelineResult that should be able to better support `PipelineResult` semantics, such as `waitUntilFinish` and `cancel`. 

Also relates to `BEAM-1000`,  `BEAM-595` and `BEAM-1050` as well as to #1438.

R: @amitsela

amitsela commented:
Sweet!
I'll review this tomorrow. Jenkins test failure seems unrelated.

amitsela commented:
@staslev before digging deep into this extensive change, which I think generally makes sense - why submit a streaming pipeline in a separate thread when Spark's driver code for streaming pipelines is asynchronous to begin with ?

amitsela commented:
I think unified-execution is indeed the essence of this PR, and that's great because it actually solves 2 or 3 open issues.
My question is about the implementation, for example: could the `ExecutorService` in `SparkRunner` be only a part of `SparkPipelineResult.BatchMode` ? and so `SparkPipelineResult.StreamingMode` is non-blocking by Spark's nature.

WDYT ?

amitsela commented:
And please add the relevant issues this PR addresses, for example: `[BEAM-123, BEAM-456] Title`.
Thanks,

staslev commented:
Technically it could, but I don't think it is as intuitive in terms of design.
The reason it (i.e. the `ExecutorService`) is in `SparkRunner#run()` is because the runner is responsible for **starting** pipeline executions, so that `SparkPipelineResult` does not have to, which it will in case the `ExecutorService` moves there. It sounds a bit off to have the `SparkPipelineResult` start executions, don't you think?

amitsela commented:
Makes sense. I'll start the review, thanks!

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/EvaluationContext.java:
```diff
@@ -195,23 +193,24 @@ public void computeOutputs() {
-  }
-
-  @Override
   public <T> AggregatorValues<T> getAggregatorValues(Aggregator<?, T> aggregator)
```
This can be removed as it is a part of `PipelineResult` and implemented by `SparkPipelineResult` now.

staslev replied:
Not sure I understand.
This method is no longer here...

amitsela replied:
`getAggregatorValues` exists in `EvaluationContext` line: 195 and it's unused.

staslev replied:
I thought I had taken care of that, must have missed it.
Take a look now.

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/EvaluationContext.java:
```diff
@@ -195,23 +193,24 @@ public void computeOutputs() {
-  @Override
-  public MetricResults metrics() {
-    throw new UnsupportedOperationException("The SparkRunner does not currently support metrics.");
+  public <T> T getAggregatorValue(String named, Class<T> resultType) {
```
Also remove, but I'll add that while this is not a part of `PipelineResult` I think it's worthing adding to `SparkPipelineResult` and possibly proposing (not as part of this PR though) to add this to `PipelineResult` as I for one, find it more useful then `getAggregatorValues`.

staslev replied:
I understand what you're saying, but I think this may introduce some less than ideal coupling between `SparkPipelineResult` and the runtime context, i.e., `SparkRuntimeContext`, in addition to the existing coupling to `EvaluationContext`.

The current design is meant to hide `SparkRuntimeContext` from `SparkPipelineResult` by exposing the needed services via `EvaluationContext`, i.e., the two overloads of the `getAggregatorValue` method. 

I was trying to avoid having stuff like `context.getRuntimeContext().getAggregatorValues(..)` in `SparkPipelineResult` since it would make `SparkPipelineResult` know about the runtime context (which I referred to as feature envy).

amitsela replied:
Well, if it's between delegating `getAggregatorValue/s` or providing `SparkRuntimeContext` - I think that it is OK to expect a context (`EvaluationContext`) to provide stuff, rather than delegate to it. WDYT ?

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/SparkPipelineStateTest.java:
```diff
@@ -0,0 +1,219 @@
+import static org.hamcrest.Matchers.nullValue;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
```
Please remove empty newlines throughout this file's imports.

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/SparkPipelineStateTest.java:
```diff
@@ -0,0 +1,219 @@
+import org.junit.rules.TestName;
+
+/**
+ * This suite tests that verifies a Beam over spark pipeline fails fast upon a failed batch and
```
I think this comment could be clarified a bit more, just seems this was constructed and improved over a couple of iterations and just needs a final touch 😉

staslev replied:
I agree, the wording seems weird...

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/SparkPipelineStateTest.java:
```diff
@@ -0,0 +1,219 @@
+
+    result.waitUntilFinish(Duration.millis(1));
+
+    assertThat(result.getState(), nullValue());
```
I'm assuming the tests pass, but I didn't notice `BatchMode` returning `null` on waitUntilFinish so how does this pass in batch ?

staslev replied:
It's taken care of in the parent class - `SparkPipelineResult#waitUntilFinish` since both flows are handled the same way now.

amitsela replied:
👍

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/SparkPipelineStateTest.java:
```diff
@@ -0,0 +1,219 @@
+    result.waitUntilFinish(Duration.millis(1));
+
+    assertThat(result.getState(), nullValue());
+  }
```
Once timeout is successfully asserted, why not cancel ? this state is generally something that didn't finish even though it timed-out so it's probably safer to "force stop" anyway.
Also maybe worth piggybacking on this to test cancel that way.

staslev replied:
I agree it may be safer to cancel, but I don't think we should mix test cases.

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/SparkPipelineStateTest.java:
```diff
@@ -0,0 +1,219 @@
+
+    assertThat(result.getState(), is(PipelineResult.State.RUNNING));
+
+    result.cancel();
```
Even better, test cancel here.. cancelling a `RUNNING` pipeline.

staslev replied:
In the scope of unit test I would prefer to not mix test cases, especially since we already have a separate test case for cancel. What do you think?

staslev commented on runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPipelineResult.java:
```diff
@@ -0,0 +1,183 @@
+ */
+public abstract class SparkPipelineResult implements PipelineResult {
+
+  protected final Future pipelineExecutionCommenced;
```
👍

staslev commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/EvaluationContext.java:
```diff
@@ -46,15 +44,13 @@
 /**
- * Evaluation context allows us to define how pipeline instructions.
+ * Evaluation context allows us to define how pipeline instructions and translate between
+ * {@code PObject<T>}s or {@code PCollection<T>}s and Ts or collections of Ts.
```
I don't mind changing it, but it's a copy paste of a previous javadoc...

staslev commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/EvaluationContext.java:
```diff
@@ -46,15 +44,13 @@
 
 /**
- * Evaluation context allows us to define how pipeline instructions.
+ * Evaluation context allows us to define how pipeline instructions and translate between
```
👍

staslev commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/EvaluationContext.java:
```diff
@@ -21,20 +21,18 @@
 import java.util.Map;
 import java.util.Set;
-import org.apache.beam.runners.spark.EvaluationResult;
+
```
👍

staslev replied:
👍

staslev commented:
Ok, I addressed the comments inline.

The newlines in the import sections keep being re-inserted whenever I reformat the code in Intellij, which I do quite often. Someday someone will have to look into it and see if this can be configured because it's damn annoying. This someone might have to be me, and the someday might need to be soon.

amitsela commented:
This is looking very well, my only comments are about removing `Aggregator`s from `EvaluationContext` now that it no longer implements `PipelineResult`. That's why you created `SparkPipelineResult`.
Thanks!

staslev commented:
I fixed the unused `getAggregatorValues` method thingy. 
This leaves us the discussion about whether `SparkPipelineResult` should know about `SparkRuntimeContext`.

staslev commented:
I have given things some further thought, and made some changes to the way `SparkPipelineResult` interacts with aggregators in an attempt to both remove the delegations you were not particularly fond of, and avoid the `x.y.z()` pattern I did not really care fore. To make it happen I also had to change peripherals.

Let me know what you think.

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/aggregators/SparkAggregators.java:
```diff
@@ -0,0 +1,99 @@
+package org.apache.beam.runners.spark.aggregators;
+
+import com.google.common.collect.ImmutableList;
+
```
Please remove empty newline.

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/aggregators/SparkAggregators.java:
```diff
@@ -0,0 +1,99 @@
+
+import java.util.Collection;
+import java.util.Map;
+
```
Here too.

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/aggregators/SparkAggregators.java:
```diff
@@ -0,0 +1,99 @@
+
+  private static <T> AggregatorValues<T> valueOf(final Accumulator<NamedAggregators> accum,
+                                                 final Aggregator<?, T> aggregator) {
+    @SuppressWarnings("unchecked") final
```
I think `final Class<T> aggValueClass...` inlinewould be more readable.

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkRuntimeContext.java:
```diff
@@ -61,12 +51,11 @@
   private transient CoderRegistry coderRegistry;
 
-  SparkRuntimeContext(Pipeline pipeline, JavaSparkContext jsc) {
+  SparkRuntimeContext(Pipeline pipeline) {
```
👍 removing the `SparkContext` is always great.

amitsela commented:
@staslev great! added a few annoying comments mostly on style 😉, ping me once addressed and I'll LGTM this. Thanks!

staslev commented:
Hopefully I got all of these empty lines this time, but you can never be sure :)

Let me know how it looks.

amitsela commented:
LGTM. I'll merge once tests pass. Thanks!

amitsela commented:
I'm assuming it is related to how you wrap thrown exceptions in `SparkPipelineResult`.

staslev commented:
@amitsela Indeed it was, should be fixed now.
